frontend/src/framework/components/ParameterListFilter/private-utils/smartNodeSelectorUtils.ts

```diff
@@ -0,0 +1,179 @@
+    treeNodeDataList.push(newNode);
+    return newNode;
+}
+
+export function findOrCreateNode(treeNodeDataList: TreeDataNode[], nodeName: string, icon?: string): TreeDataNode {
+    const existingNode = treeNodeDataList.find((node) => node.name === nodeName);
+    if (existingNode) {
+        return existingNode;
+    }
+
+    return createAndAddNode(treeNodeDataList, nodeName, icon);
+}
+
+export function addParameterNameAndGroupToTreeDataNodeList(
+    treeNodeDataList: TreeDataNode[],
+    parameter: Parameter,
+    icon?: string
+): void {
+    // Parameter Name
+    const nameParentNode = findOrCreateNode(treeNodeDataList, ParameterParentNodeNames.NAME, icon);
+    if (!nameParentNode.children) {
+        nameParentNode.children = [];
+    }
+    findOrCreateNode(nameParentNode.children, parameter.name);
+
+    // Parameter Group
+    if (parameter.groupName) {
+        const groupParentNode = findOrCreateNode(treeNodeDataList, ParameterParentNodeNames.GROUP, icon);
+        if (!groupParentNode.children) {
+            groupParentNode.children = [];
+        }
+        findOrCreateNode(groupParentNode.children, parameter.groupName);
+    }
+}
+
+export function createTreeDataNodeListFromParameters(
+    parameters: Parameter[],
+    checkIcon?: string,
+    parentIcon?: string
+): TreeDataNode[] {
+    if (parameters.length === 0) {
+        return [];
+    }
+
+    const treeDataNodeList: TreeDataNode[] = [];
+
+    const hasContinuousParameter = parameters.some((parameter) => parameter.type === ParameterType.CONTINUOUS);
+
+    // Node for boolean/state properties on top level
+    createAndAddNode(treeDataNodeList, ParameterParentNodeNames.CONTINUOUS, checkIcon);
+    createAndAddNode(treeDataNodeList, ParameterParentNodeNames.DISCRETE, checkIcon);
+    createAndAddNode(treeDataNodeList, ParameterParentNodeNames.IS_CONSTANT, checkIcon);
+    createAndAddNode(treeDataNodeList, ParameterParentNodeNames.IS_NONCONSTANT, checkIcon);
+    if (hasContinuousParameter) {
+        createAndAddNode(treeDataNodeList, ParameterParentNodeNames.IS_LOGARITHMIC, checkIcon);
+        createAndAddNode(treeDataNodeList, ParameterParentNodeNames.IS_LINEAR, checkIcon);
+    }
+
+    // Add name and group for parameters
+    for (const parameter of parameters) {
+        addParameterNameAndGroupToTreeDataNodeList(treeDataNodeList, parameter, parentIcon);
+    }
+
+    return treeDataNodeList;
+}
+
+export function getParametersMatchingSelectedNodes(parameters: Parameter[], selectedNodes: string[]): Parameter[] {
+    // No selection implies no filtering
+    if (selectedNodes.length === 0) {
+        return parameters;
+    }
+
+    const delimiter = ":";
+    const findSelectedParameterPropertiesFromName = (propertyName: string): string[] => {
+        return selectedNodes
+            .filter((node) => node.split(delimiter, 1)[0] === propertyName)
+            .map((node) => node.split(delimiter, 2)[1]);
+    };
```

Comment:
Consider moving out of this function and make it a regular function (for debugging). Also, I find the name a bit confusing.

Reply:
Done. 

For scoped functionality I sometimes like internal functions. And this is an example where i think arrow-functions come to hand - where it has the `this` content of the parent function, i.e. less arguments.